File: vb_nc_check.py
# pip install pywin32
import win32com.client

def run_normcad_check(norm, task_name, unit, vars_progid, vars_values: dict, conds=None):
    # 1) Create API object
    nc = win32com.client.Dispatch("ncApi.Report")  # :contentReference[oaicite:1]{index=1}

    # 2) Select module/task
    nc.Norm = norm
    nc.TaskName = task_name
    nc.Unit = unit

    nc.ClcLoadNorm  # иногда как свойство, иногда как метод (зависит от COM обертки)
    try:
        nc.ClcLoadNorm()
    except TypeError:
        pass

    # 3) Create Vars object of конкретного расчетного модуля
    Vars = win32com.client.Dispatch(vars_progid)  # пример: "NC_219532362921554E02.Vars" :contentReference[oaicite:2]{index=2}

    # 4) Fill input variables (имена полей берешь из “Создать VB проект”)
    for k, v in vars_values.items():
        setattr(Vars, k, v)

    nc.SetVars(Vars)  # :contentReference[oaicite:3]{index=3}

    # 5) Conditions (если надо)
    if conds:
        nc.SetConds(conds)  # :contentReference[oaicite:4]{index=4}

    # 6) Load data/conds and calculate
    nc.ClcLoadData()
    nc.ClcLoadConds()
    nc.ClcCalc()  # 

    max_util = float(nc.MaxResult)  # 
    return {"passed": max_util <= 1.0, "max_util": max_util}

# пример вызова (значения norm/task/unit/vars_progid и имена Vars.* — из сгенерированного VB-проекта)
# res = run_normcad_check("СП 64.13330...", "Балка", "1", "NC_....Vars", {"L": 5800, "b": 200, "h": 220, "q": 2.5})
# print(res)

File: vb_nc_check.vb
' VB/VBA skeleton for running one check
Option Explicit

Public Type CheckResult
    Passed As Boolean
    MaxUtil As Double
    Message As String
End Type

Public Function CheckTimberBeam_SP64(ByVal L_mm As Double, ByVal b_mm As Double, ByVal h_mm As Double, _
                                    ByVal q_kN_per_m As Double, ByVal materialName As String) As CheckResult
    Dim r As CheckResult
    On Error GoTo EH

    Dim ncApiR As Object
    Set ncApiR = CreateObject("ncApi.Report") ' NCAPI.dll 

    ' 1) Choose module / task
    ncApiR.Norm = "СП 64.13330.2017 Деревянные конструкции"  ' пример: уточни по своему модулю
    ncApiR.TaskName = "Балка"                                ' взять из сгенерированного проекта
    ncApiR.Unit = "1"                                        ' взять из сгенерированного проекта

    ncApiR.ClcLoadNorm

    ' 2) Vars object (ProgID берёшь из сгенерированного VB проекта)
    Dim Vars As Object
    Set Vars = CreateObject("NC_XXXXXXXXXXXXXXX.Vars")

    ' 3) Assign input vars (имена полей — из сгенерированного проекта!)
    Vars.L = L_mm
    Vars.b = b_mm
    Vars.h = h_mm
    Vars.q = q_kN_per_m
    Vars.Material = materialName

    ncApiR.SetVars Vars

    ' 4) Conditions array (если нужно) — тоже видно в сгенерированном проекте
    Dim Conds(0 To 0) As String
    Conds(0) = "..."   ' например: схема опирания / длительность нагрузки / класс эксплуатации
    ncApiR.SetConds Conds

    ncApiR.ClcLoadData
    ncApiR.ClcLoadConds
    ncApiR.ClcCalc

    r.MaxUtil = ncApiR.MaxResult  ' 
    r.Passed = (r.MaxUtil <= 1#)

    If r.Passed Then
        r.Message = "OK"
    Else
        r.Message = "FAIL: Max utilization = " & CStr(r.MaxUtil)
    End If

    CheckTimberBeam_SP64 = r
    Exit Function

EH:
    r.Passed = False
    r.MaxUtil = 9999
    r.Message = "ERROR: " & Err.Description
    CheckTimberBeam_SP64 = r
End Function
